pkgs/development/interpreters/python/editable.nix

```diff
@@ -0,0 +1,99 @@
+{
+  buildPythonPackage,
+  lib,
+  hatchling,
+  tomli-w,
+}:
+{
+  pname,
+  version,
+
+  # Editable root as string.
+  # Environment variables will be expanded at runtime using os.path.expandvars.
+  root,
+
+  # Arguments passed on verbatim to buildPythonPackage
+  derivationArgs ? { },
+
+  # Python dependencies
+  dependencies ? [ ],
+  optional-dependencies ? { },
+
+  # PEP-518 build-system https://peps.python.org/pep-518
+  build-system ? [ ],
+
+  # PEP-621 entry points https://peps.python.org/pep-0621/#entry-points
+  scripts ? { },
+  gui-scripts ? { },
+  entry-points ? { },
+
+  passthru ? { },
+  meta ? { },
+}:
+
+# Create a PEP-660 (https://peps.python.org/pep-0660/) editable package pointing to an impure location outside the Nix store.
+# The primary use case of this function is to enable local development workflows where the local package is installed into a virtualenv-like environment using withPackages.
+
+assert lib.isString root;
+let
+  # In editable mode build-system's are considered to be runtime dependencies.
+  dependencies' = dependencies ++ build-system;
+
+  pyproject = {
+    # PEP-621 project table
+    project = {
+      name = pname;
+      inherit
+        version
+        scripts
+        gui-scripts
+        entry-points
+        ;
+      dependencies = map lib.getName dependencies';
+      optional-dependencies = lib.mapAttrs (_: lib.getName) optional-dependencies;
+    };
+
+    # Allow empty package
+    tool.hatch.build.targets.wheel.bypass-selection = true;
+
+    # Include our editable pointer file in build
+    tool.hatch.build.targets.wheel.force-include."_${pname}.pth" = "_${pname}.pth";
+
+    # Build editable package using hatchling
+    build-system = {
+      requires = [ "hatchling" ];
+      build-backend = "hatchling.build";
+    };
+  };
+
+in
+buildPythonPackage (
+  {
+    inherit
+      pname
+      version
+      optional-dependencies
+      passthru
+      meta
+      ;
+    dependencies = dependencies';
+
+    pyproject = true;
+
+    unpackPhase = ''
+      python -c "import json, tomli_w; print(tomli_w.dumps(json.load(open('$pyprojectContentsPath'))))" > pyproject.toml
+      echo 'import os.path, sys; sys.path.insert(0, os.path.expandvars("${root}"))' > _${pname}.pth
```

Comment:
@adisbladis should this be `sys.path.append` instead of prepend.

the issue we running into is now if someone prepends to PYTHONPATH in shell, those entries no longer first.

thoughts ?

Reply:
It's deliberately been changed to append, as pretending would break semantics of editables see https://github.com/NixOS/nixpkgs/pull/339228#discussion_r1741901135

Reply:
This makes overrides via PYTHONPATH impossible.  

As for editable semantics,  https://setuptools.pypa.io/en/latest/userguide/development_mode.html

mentions no guarantee about loading order

Reply:
```
python  -m venv venv
pip install -e .
python -c "import sys; print(sys.path)"
```
you can see that editable install is last in path ( appended ), and overrides why PYTHONPATH still possible

we have same behavior

Reply:
This is about nixpkgs, so if you get the same behavior with https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/sitecustomize.py active, that would be more convincing.

It makes overriding via pythonpath impossible for packages which one explicitly added as editables to be precise, far less drastic in scope.

PYTHONPATH is arguably subtly broken due to the way nixpkgs does input propagation in Python packages and handles the fallout, an issue which has been discussed at length so I am not sure whether a 4 month old, merged PR is the best way to do so again.